File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/drawable/TeamspaceIconDrawable.java
package com.dashlane.ui.drawable;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;

import com.dashlane.R;
import com.dashlane.util.graphics.TextDrawable;

import java.util.Locale;

import androidx.core.content.res.ResourcesCompat;



public class TeamspaceIconDrawable extends TextDrawable {

    private static final float TEXT_SIZE_FACTOR = 0.5f;
    private static final float BORDER_SIZE_RATIO = 2 / 24f; 

    private final Path mPath = new Path();
    private final Paint mPaintBorder = new Paint();
    private final Rect mTempRect = new Rect();
    private int mSize = -1;

    private TeamspaceIconDrawable(Context context, String text, int color) {
        super(text,
              Color.WHITE,
              color,
              ResourcesCompat.getFont(context, R.font.gt_walsheim_pro_bold));
        mPaintBorder.setColor(Color.BLACK);
        mPaintBorder.setAlpha(Math.round(0.24f * 256));
        mPaintBorder.setAntiAlias(true);
        mPaintBorder.setStyle(Paint.Style.STROKE);
    }

    public static TeamspaceIconDrawable newInstance(Context context, String label, int color) {
        String icoLabel;
        if (label == null || label.length() == 0) {
            icoLabel = null;
        } else {
            icoLabel = label.substring(0, 1).toUpperCase(Locale.US);
        }
        return new TeamspaceIconDrawable(context, icoLabel, color);
    }

    public void setSize(int size) {
        mSize = size;
        invalidateSelf();
    }

    @Override
    protected void drawBackground(Canvas canvas, Rect bounds, Paint paint) {
        fillBackgroundPath(bounds, mPath);
        canvas.drawPath(mPath, paint);

        mTempRect.set(bounds);
        int borderSize = Math.round((float) bounds.width() * BORDER_SIZE_RATIO);

        float halfBorderSize = borderSize / 2f;
        mTempRect.bottom -= halfBorderSize;
        mTempRect.top += halfBorderSize;
        mTempRect.right -= halfBorderSize;
        mTempRect.left += halfBorderSize;
        mPaintBorder.setStrokeWidth(borderSize);

        fillBackgroundPath(mTempRect, mPath);
        canvas.drawPath(mPath, mPaintBorder);
    }

    @Override
    protected float getTextSizeFactor() {
        return TEXT_SIZE_FACTOR;
    }

    private void fillBackgroundPath(Rect bounds, Path path) {
        int width = bounds.width();
        int height = bounds.height();
        float thirdWidth = width / 3f;
        float thirdHeight = height / 3f;
        path.reset();
        path.moveTo(bounds.left + thirdWidth, bounds.top);
        path.lineTo(bounds.right, bounds.top);
        path.lineTo(bounds.right, bounds.bottom - thirdHeight);
        path.lineTo(bounds.right - thirdWidth, bounds.bottom);
        path.lineTo(bounds.left, bounds.bottom);
        path.lineTo(bounds.left, bounds.top + thirdHeight);
        path.close();
    }

    @Override
    public int getIntrinsicWidth() {
        return mSize;
    }

    @Override
    public int getIntrinsicHeight() {
        return mSize;
    }
}

File: dashlane-project/modules/libraries/teamspace/src/main/java/com/dashlane/teamspaces/adapter/TeamspaceIconDrawable.kt
package com.dashlane.teamspaces.adapter

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Rect
import androidx.annotation.ColorInt
import androidx.core.content.res.ResourcesCompat
import com.dashlane.ui.R
import com.dashlane.util.graphics.TextDrawable

class TeamspaceIconDrawable(
    context: Context,
    displayLetter: Char,
    @ColorInt color: Int
) : TextDrawable(
    displayLetter.toString(),
    Color.WHITE,
    color,
    ResourcesCompat.getFont(context, R.font.gt_walsheim_pro_bold)
) {
    private val path = Path()

    private val paintBorder = Paint()
    private val tempRect = Rect()
    var size = -1
        set(value) {
            field = value
            invalidateSelf()
        }

    init {
        paintBorder.color = Color.BLACK
        paintBorder.alpha = Math.round(0.24f * 256)
        paintBorder.isAntiAlias = true
        paintBorder.style = Paint.Style.STROKE
    }

    override fun drawBackground(canvas: Canvas, bounds: Rect, paint: Paint) {
        fillBackgroundPath(bounds, path)
        canvas.drawPath(path, paint)
        tempRect.set(bounds)
        val borderSize = Math.round(bounds.width().toFloat() * BORDER_SIZE_RATIO)
        val halfBorderSize = borderSize / 2f
        tempRect.bottom = (tempRect.bottom - halfBorderSize).toInt()
        tempRect.top = (tempRect.top + halfBorderSize).toInt()
        tempRect.right = (tempRect.right - halfBorderSize).toInt()
        tempRect.left = (tempRect.left + halfBorderSize).toInt()
        paintBorder.strokeWidth = borderSize.toFloat()
        fillBackgroundPath(tempRect, path)
        canvas.drawPath(path, paintBorder)
    }

    override fun getTextSizeFactor(): Float {
        return TEXT_SIZE_FACTOR
    }

    private fun fillBackgroundPath(bounds: Rect, fillPath: Path) {
        val width = bounds.width()
        val height = bounds.height()
        val thirdWidth = width / 3f
        val thirdHeight = height / 3f
        fillPath.reset()
        fillPath.moveTo(bounds.left + thirdWidth, bounds.top.toFloat())
        fillPath.lineTo(bounds.right.toFloat(), bounds.top.toFloat())
        fillPath.lineTo(bounds.right.toFloat(), bounds.bottom - thirdHeight)
        fillPath.lineTo(bounds.right - thirdWidth, bounds.bottom.toFloat())
        fillPath.lineTo(bounds.left.toFloat(), bounds.bottom.toFloat())
        fillPath.lineTo(bounds.left.toFloat(), bounds.top + thirdHeight)
        fillPath.close()
    }

    override fun getIntrinsicWidth(): Int {
        return size
    }

    override fun getIntrinsicHeight(): Int {
        return size
    }

    companion object {
        private const val TEXT_SIZE_FACTOR = 0.5f
        private const val BORDER_SIZE_RATIO = 2 / 24f 
    }
}
